corgi/core/migrations/0075_install_stored_proc.py

```diff
@@ -0,0 +1,265 @@
+'[[:digit:]]', ''), 1) + 1);
+            END LOOP;
+            WHILE (str2 IS NOT NULL AND REGEXP_REPLACE(str2,
+'[[:digit:]]', '') IS NULL) LOOP
+                str2 := SUBSTR(str2, INSTR(str2, REGEXP_REPLACE(str2,
+'[[:digit:]]', ''), 1) + 1);
+            END LOOP;
+            isnum := 1;
+        ELSE
+            WHILE (str1 IS NOT NULL AND REGEXP_REPLACE(str1,
+'[[:alpha:]]', '') IS NULL) LOOP
+                str1 := SUBSTR(str1, INSTR(str1, REGEXP_REPLACE(str1,
+'[[:alpha:]]', ''), 1) + 1);
+            END LOOP;
+            WHILE (str2 IS NOT NULL AND REGEXP_REPLACE(str2,
+'[[:alpha:]]', '') IS NULL) LOOP
+                str2 := SUBSTR(str2, INSTR(str2, REGEXP_REPLACE(str2,
+'[[:alpha:]]', ''), 1) + 1);
+            END LOOP;
+            isnum := 0;
+        END IF;
+
+        oldch1 := SUBSTR(str1, LENGTH(str1));
+        str1 := SUBSTR(str1, 1, LENGTH(str1) - 1);
+        oldch2 := SUBSTR(str2, LENGTH(str2));
+        str2 := SUBSTR(str2, 1, LENGTH(str2) - 1);
+
+        IF two IS NULL THEN
+            RETURN (isnum + 1) / 2;
+        END IF;
+
+        IF one = str1 THEN
+            RETURN (isnum - 1) / 2;
+        END IF;
+
+        IF isnum = 1 THEN
+            IF LENGTH(one) > LENGTH(str2) THEN
+                RETURN 1;
+            END IF;
+
+            IF LENGTH(str2) > LENGTH(one) THEN
+                RETURN -1;
+            END IF;
+        END IF;
+
+        rc := CASE WHEN one < str2 THEN -1 ELSE 1 END;
+
+        IF rc <> 0 THEN
+            RETURN rc;
+        END IF;
+
+        one := str1 || oldch1;
+        two := str2 || oldch2;
+    END LOOP;
+
+    -- identical though segment seperating was different
+    IF one IS NULL AND two IS NULL THEN
+        RETURN 0;
+    END IF;
+
+    -- whichever version still has characters left is deemed the winner
+    IF one IS NULL THEN
+        RETURN -1;
+    ELSE
+        RETURN 1;
+    END IF;
+END;
+ $$
+            """
+        ),  # noqa: E501
+        migrations.RunSQL("DROP FUNCTION if exists get_latest_component;"),
+        migrations.RunSQL(
+            """
+            CREATE FUNCTION get_latest_component(model_type text, ps_ofuri text, component_ns text, component_name text) RETURNS text LANGUAGE PLPGSQL AS $$  
+  DECLARE
+    COMPONENT RECORD;
+    latest_uuid text;
+    latest_version text;
+    compare_versions int;
+  BEGIN
+    IF model_type = 'ProductStream' THEN
+      FOR COMPONENT IN select *, row_number() over () from core_component
+           INNER JOIN "core_component_productstreams"
+           ON ("core_component"."uuid" = "core_component_productstreams"."component_id")
+           INNER JOIN "core_productstream"
+           ON ("core_component_productstreams"."productstream_id" = "core_productstream"."uuid")
+           where core_component.namespace=component_ns and core_component.name=component_name and ( (core_component.type='RPM' and core_component.arch='src') or (core_component.type='OCI' and core_component.arch='noarch'))
```

Comment:
I think we might not need this part because that's already done as part of the root_components filter, which we are still using previous to using this stored procedure. That also gives us the flexibility to use this stored procedure to filter things which are not root components.

Reply:
sounds reasonable ;) will remove to make more generic

Reply:
Maybe it does make sense to include this condition in the stored procedure, perhaps adding RPMMOD as well. The current root component condition includes Github components for managed serviced, which are unlikely to have version matching RPM schematics.

Reply:
yes I need to refresh it as there has been some changes - will do